File: 0567-permutation-in-string/0567-permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        if(s1.length() > s2.length())
            return false;
        
        HashMap<Character, Integer> M = new HashMap<>();
        for(int i=0; i<s1.length(); i++){
            char c = s1.charAt(i);
            var cur_val = M.computeIfAbsent(c, k -> 0);
            M.put(c, cur_val+1);
        }
        int cnt = M.size();
        
        for(int i=0; i<s1.length(); i++){
            char cur_char = s2.charAt(i);
            if(M.containsKey(cur_char)){
                int prev_cnt = M.put(cur_char, M.get(cur_char)-1);
                if(prev_cnt == 1)
                    cnt--;
            }
        }
        
        int l = 1, r = s1.length();
        while(r<s2.length()){
            if(cnt==0)
                return true;
            
            var l_char = s2.charAt(l-1);
            var r_char = s2.charAt(r);
            if(M.containsKey(l_char)){
                int prev_cnt = M.put(l_char, M.get(l_char)+1);
                if(prev_cnt == 0)
                    cnt++;
            }
            if(M.containsKey(r_char)){
                int prev_cnt = M.put(r_char, M.get(r_char)-1);
                if(prev_cnt == 1)
                    cnt--;
            }
            
            l++;
            r++;
        }
        
        return cnt==0;
    }
}

File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if(s1.length() > s2.length())
            return false;
        
        vector<int> count(26);
        int non_zero = 0;
        for(char c : s1){
            if(count[c-'a']==0) 
                non_zero++;
            count[c-'a']++;
        }
        
        for(int i=0; i<s1.size(); i++){
            char c = s2[i];
            count[c-'a']--;
            if(count[c-'a']==0)
                non_zero--;
            else if(count[c-'a']==-1)
                non_zero++;
        }
        
        if(non_zero == 0)
            return true;
        
        int l = 0, r = s1.size();
        while(r<s2.size()){
            char r_c = s2[r], l_c = s2[l];
            count[r_c-'a']--;
            if(count[r_c-'a']==0)
                non_zero--;
            else if(count[r_c-'a']==-1)
                non_zero++;
                
            count[l_c-'a']++;
            if(count[l_c-'a']==0)
                non_zero--;
            else if(count[l_c-'a']==1)
                non_zero++;
                
            if(non_zero==0)
                return true;
            
            l++;
            r++;
        }
        
        return false;
        
        
    }
};